interface/imatcopy.c

```diff
@@ -120,71 +120,71 @@ void CNAME( enum CBLAS_ORDER CORDER, enum CBLAS_TRANSPOSE CTRANS, blasint crows,
     		BLASFUNC(xerbla)(ERROR_NAME, &info, sizeof(ERROR_NAME));
     		return;
   	}
+
 #ifdef NEW_IMATCOPY
-    if ( *lda == *ldb && *rows == *cols) {
+    if ( *lda == *ldb ) {
         if ( order == BlasColMajor )
         {
             if ( trans == BlasNoTrans )
             {
                 IMATCOPY_K_CN(*rows, *cols, *alpha, a, *lda );
+                return;
             }
-            else
+            else if ( *rows == *cols )
             {
                 IMATCOPY_K_CT(*rows, *cols, *alpha, a, *lda );
+                return;
             }
         }
         else
         {
             if ( trans == BlasNoTrans )
             {
                 IMATCOPY_K_RN(*rows, *cols, *alpha, a, *lda );
+                return;
             }
-            else
+            else if ( *rows == *cols )
             {
                 IMATCOPY_K_RT(*rows, *cols, *alpha, a, *lda );
+                return;
             }
         }
-        return; 
     }
-
 #endif
 
-	if ( *lda >  *ldb )
-		msize = (size_t)(*lda) * (*ldb)  * sizeof(FLOAT);
-	else
-		msize = (size_t)(*ldb) * (*ldb)  * sizeof(FLOAT);
+	msize = (size_t)(*rows) * (*cols) * sizeof(FLOAT);
```

Comment:
Also, what if `*ldb` is larger than the corresponding `*rows` or `*cols`? Wouldn't the allocated memory be too small with the new logic?

Reply:
IIUC, the original code over-allocates memory if `*lda` is larger than the actual number of `*rows` or `*cols` it corresponds to.
The proposed change under-allocates if `*ldb` is larger than the actual number of `*rows` or `*cols` it corresponds to.

Reply:
The following could be correct:
```
if ( (trans == BlasNoTrans) != (order == BlasColMajor) )
  // output is row major
  msize = (size_t)(*rows) * (*ldb) * sizeof(FLOAT);
else
  // output is col major
  msize = (size_t)(*cols) * (*ldb) * sizeof(FLOAT);
```

But please double check.

Reply:
Updated that snippet twice forth and back. That's a bigger head-scratcher than I'd like to admit.
But I think the last iteration is correct now. (As was the first. But I didn't believe in myself. 😉)

Reply:
I find it most helpful to think about how many entries are copied from `a` to `b` and then back again to `a`: the answer is simply `rows * cols`, regardless of the input (`lda`) and output (`ldb`) strides. (As a motivating example, consider transposing just a `10 x 10` block of a `10000 x 10000` matrix.)

Note, importantly, that `ldb` doesn't really have anything to do with `b` (maybe this is the cause of the confusion): `ldb` is the stride for the output matrix, which is stored in `a`, specifying where/how to write the data coming from `b`; while `b` just needs to hold that data, in principle with any stride it likes (as a purely internal implementation detail) -- but ideally stored as compactly as possible. From this point of view, it's clear that `msize` (= size of `b`) should not depend on `ldb` at all. (A more suggestive renaming of variables might read: `a` -> `ab` (as in the MKL documentation), `b` -> `tmp`, and `msize` -> `tmpsize`.)

The old code breaks precisely when `trans = BlasNoTrans` (or equivalent) because then `lda` and `ldb` both only give you information on the same dimension (either rows or columns, but not both), as in the opening test case in #4038. (It can be "fixed" by allocating `lda * ldb` entries instead, but this would not be optimal since `lda * ldb >= rows * cols`.)

Reply:
I might be misreading the MKL documentation. But afaict, `*ldb` is the stride that the function caller selects for the output matrix. `msize` is the memory in bytes that is allocated for that output.
As an example, assume a column-major output matrix `M` with 2×3 elements and a user-selected stride of 10. In that case, `M[10]` should hold the first value in the second value. If the memory allocated for `M` is only `2*3`, that would be out of bounds.

If this is correct, please check if it is really possible to replace `ldb` in the function calls further down in this file.

Reply:
Oops. I was reading a page that was a bit misleading (I don't remember where I found it). The documentation here is much clearer:
https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-0/imatcopy.html

You are completely right. `a` is the input *and* output buffer. `b` is just a temporary intermediate. Having it exactly match `*rows`*`*cols` is large enough.
Your proposed code looks good to me.

I'm sorry for having caused that confusion.